GH-1772 Add RetryTopicNamesProvider

apolci opened the pull request:
Resolves #1772 

Add provider to handle the names used on retry/dl topics (topic name, groupId, etc.). This can be extended to customize the behaviour in addition to the simple suffix parameter.

Currently customizing the default requires to register a bean  implementing RetryTopicNamesProviderFactory interface, with name "internalRetryTopicNamesProviderFactory".

garyrussell commented on spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurer.java:
```diff
@@ -221,14 +221,18 @@
 		this.listenerContainerFactoryConfigurer = listenerContainerFactoryConfigurer;
 		this.beanFactory = beanFactory;
+		this.retryTopicNamesProviderFactory = retryTopicNamesProviderFactory;
 	}
```
Since this is a public API and we have released 2.7.0 already, we need to keep the old constructor (and deprecate it), populating the new factory with the default.

It is not necessary to do this on the non-public APIs.

garyrussell commented on spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicNamesProviderFactory.java:
```diff
@@ -0,0 +1,42 @@
+
+		String getGroup(MethodKafkaListenerEndpoint<?, ?> endpoint);
+
+		String getTopicName(String topic);
```
These need javadocs; blank line before first method and between `}` below.

garyrussell commented on spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurer.java:
```diff
@@ -221,14 +221,18 @@
+								ListenerContainerFactoryConfigurer listenerContainerFactoryConfigurer,
+								BeanFactory beanFactory,
+								RetryTopicNamesProviderFactory retryTopicNamesProviderFactory) {
 		this.destinationTopicProcessor = destinationTopicProcessor;
```
Our convention for multi-line signatures is to add a blank line before the code.

garyrussell commented on spring-kafka/src/main/java/org/springframework/kafka/retrytopic/DefaultRetryTopicNamesProviderFactory.java:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	public static class DefaultRetryTopicNamesProvider implements RetryTopicNamesProvider {
+		private final Suffixer suffixer;
```
blank line above; and line # 28 and between last `}` at the end of the file.

garyrussell commented on spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicConfigurer.java:
```diff
@@ -494,4 +501,6 @@ public void logMessage(Object message) {
 		}
 	}
+
+
```
remove one of these blank lines.

apolci replied:
Will apply requested changes.

garyrussell reviewed (changes requested):
@apolci Thanks for the contribution! Just some minor issues.

LGTM otherwise. @tomazfernandes any objections?

tomazfernandes commented on spring-kafka/src/main/java/org/springframework/kafka/retrytopic/RetryTopicNamesProviderFactory.java:
```diff
@@ -0,0 +1,42 @@
+
+	RetryTopicNamesProvider createRetryTopicNamesProvider(DestinationTopic.Properties properties);
+
+	interface RetryTopicNamesProvider {
```
I don't think I remember having ever seen an interface inside another. One more trick for the book, thanks!

tomazfernandes commented:
No objections whatsoever, looks good! I've just added a comment about the default implementation class name, but it's a matter of personal taste more than anything.

tomazfernandes commented:
Also @garyrussell do you think this is a feature worth documenting, or should we leave it for more advanced users looking for extension points? It seems like a pretty useful customisation, and a not so hard one to implement.

garyrussell commented:
@tomazfernandes Yes, we should add a few words to the docs; perhaps you can guide @apolci where it should go.

Thanks.

tomazfernandes commented:
@apolci it'd be great if you could add this feature to the documentation. We already have a section on Topic Naming: https://docs.spring.io/spring-kafka/reference/html/#topic-naming, 4.4.5, so I think it'd be a matter of adding it after the existing  text, before 4.4.6. 

What do you think?

It's in this file:
https://github.com/spring-projects/spring-kafka/blob/main/spring-kafka-docs/src/main/asciidoc/retrytopic.adoc

apolci commented:
> @apolci it'd be great if you could add this feature to the documentation...

Yes sure, I'll try to do it tomorrow.

I'm not 100% sure of how it can be configured at the moment. I followed the handling of the other constructor parameters of RetryTopicConfigurer regarding the initialization in RetryTopicBootstrapper.
This means, if I'm not wrong, that a bean named "internalRetryTopicNamesProviderFactory" need to be present in the context in order to override the default handling. Do you think this is the correct way to do that?

tomazfernandes commented:
> > @apolci it'd be great if you could add this feature to the documentation...
> 
> Yes sure, I'll try to do it tomorrow.
> 
> I'm not 100% sure of how it can be configured at the moment. I followed the handling of the other constructor parameters of RetryTopicConfigurer regarding the initialization in RetryTopicBootstrapper.
> This means, if I'm not wrong, that a bean named "internalRetryTopicNamesProviderFactory" need to be present in the context in order to override the default handling. Do you think this is the correct way to do that?

Yep, that's what I had in mind. But maybe since this interface is really about the retry topic logic and thus there's little chance the user will have more than one bean implementing it for a different purpose, we might be a little more lenient and try to find the bean by type in the bootstrapper logic, and refraining to register the bean if there's already one there, so that the user doesn't have to know the exact bean name.

I'm ok with both approaches.

apolci commented:
> ...
> I'm ok with both approaches.

I implemented the second approach (search bean by class) and added the documentation. Sorry for the many commits, should I squash them?

garyrussell commented:
No; best to keep commits separate; we will squash during merge.

Thanks!

apolci commented:
Sorry, I still see 1 change request pending, but I think I resolved all of them. I just want to be sure I'm not missing something

garyrussell commented:
That's from the previous review; I just haven't gotten around to re-reviewing (and we have a day off tomorrow, so it will probably be next week now).

Thanks for your patience.

tomazfernandes commented on spring-kafka-docs/src/main/asciidoc/retrytopic.adoc:
```diff
@@ -429,6 +429,50 @@ public RetryTopicConfiguration myRetryTopic(KafkaTemplate<String, MyPojo> templa
+----
+====
+
+As an example the following implementation, in addition to the standard suffix, add a prefix to retry/dl topics names:
```
I think it should be "adds a prefix", shouldn't it?

apolci replied:
Yes, thank you 👍

tomazfernandes commented:
Apart from a tiny typo I just commented on looks good to me!

garyrussell commented:
@apolci Thanks for the contribution (and @tomazfernandes thanks for reviewing).
